### PR TITLE
Engine - Preservation of Original YAML Asset

### PR DESCRIPTION
**Pull request by NahuFigueroa97**
|Related issue|
|---|
|#17847|

## Description

Currently, when saving an asset in the Wazuh engine, we convert the asset from YAML to JSON and store it. However, this conversion process results in information loss and a discrepancy between the original asset and its visual representation. To address these issues, we propose a solution that involves storing both the converted JSON asset and the original YAML asset. By doing so, we can provide users with the unaltered asset when requested via the CLI or API, eliminating any transformation-related discrepancies.

## Requeriments

The asset conversion process from YAML to JSON and vice versa presents the following challenges:

- Loss of information: During the initial conversion, YAML comments are discarded, depriving users and administrators of valuable contextual information.

- Inaccurate visual representation: The expanded arrays in the JSON representation deviate from the original YAML input, making it difficult for users to comprehend the asset structure accurately.

## Tasks

 - [x] Modify the way of creating, updating and reading from the catalog.
 - [x] Add a flag to the get parameter of the catalog command to get the original yaml
 - [x] Modify the integration script so that it loads the original file
 - [x] Unit Test

### CODE REVIEW

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 1]
```diff
@@ -177,7 +177,16 @@ std::optional<base::Error> Catalog::postResource(const Resource& collection, con
                                         std::get<base::Error>(formatResult).message)};
     }
 
-    const auto contentJson = std::get<json::Json>(formatResult);
+    // Wraps the content in both json and yml
+    auto wrappedContent = json::Json {R"({})"};
```
Comment (JavierBejMen) [inline src/engine/source/api/src/catalog/catalog.cpp:181]: Please don't use the parser constructor, use the default initializer and setObject() method.

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 2]
```diff
@@ -164,6 +167,17 @@ std::optional<base::Error> Catalog::postResource(const Resource& collection, con
...
+    Format contentFormat;
+    try
+    {
+        auto tmp = json::Json {content.c_str()};
+        contentFormat = Format::json;
+    }
+    catch (const std::exception& e)
+    {
+        contentFormat = Format::yaml;
+    }
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:179]: How can you be sure that if it is not json it is a valid yaml? You shouldn't auto deduce the type, it's information you already have.

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 3]
```diff
@@ -258,6 +274,17 @@ std::optional<base::Error> Catalog::putResource(const Resource& item, const std:
...
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:286]: How can you be sure that if it is not json it is a valid yaml? You shouldn't auto deduce the type, it's information you already have.

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 4]
```diff
@@ -323,33 +351,66 @@ std::optional<base::Error> Catalog::putResource(const Resource& item, const std:
...
+    using Format = ::com::wazuh::api::engine::catalog::ResourceFormat;
+
+    const auto handleError = [](const std::string& errorMessage)
     {
-        return base::Error {fmt::format("Content '{}' could not be obtained from store: {}",
-                                        resource.m_name.fullName(),
-                                        std::get<base::Error>(storeResult).message)};
-    }
+        return base::Error {errorMessage};
+    };
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:360]: This is unnecessary, adds complexity to the code and does nothing.

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 5]
```diff
@@ -323,33 +351,66 @@ std::optional<base::Error> Catalog::putResource(const Resource& item, const std:
...
+            return handleError(fmt::format("JSON object could not be created from '{} {}': {}",
+                                           Resource::formatToStr(resource.m_format),
+                                           resource.m_name.fullName(),
+                                           std::get<base::Error>(formatResult).message));
+        }
+
+        return std::get<std::string>(formatResult);
+    };
+
+    const auto getFormattedContent = [&](const json::Json& content) -> std::variant<std::string, base::Error>
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:382]: same

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 6]
```diff
@@ -323,33 +351,66 @@ std::optional<base::Error> Catalog::putResource(const Resource& item, const std:
...
+            if (content.exists("/original"))
+            {
+                json::Json jsonCopy = content; // Create a copy to modify
+                jsonCopy.erase("/original");
+                return formatContent(jsonCopy.getJson("/json").value());
+            }
+
+            return formatContent(content);
+        }
+        else if (Resource::formatToStr(Format::json) == content.getString("/format").value())
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:395]: else is not necessary

File: src/engine/source/builder/policy.hpp [hunk 7]
```diff
@@ -179,15 +179,13 @@ class Policy
                     throw std::runtime_error("Integration name is not a string");
                 }
 
-                auto integrationDef = storeRead->get(integration.getString().value());
-                if (std::holds_alternative<base::Error>(integrationDef))
+                auto jsonObject = store::get(storeRead, integration.getString().value());
```
Comment (juliancnn) [inline src/engine/source/builder/policy.hpp:182]: restore the original name, its de integration definition

File: src/engine/source/builder/policy.hpp [hunk 8]
```diff
@@ -271,7 +269,7 @@ class Policy
 
     static std::unordered_map<std::string, std::vector<std::shared_ptr<Asset>>>
     getManifestAssets(const json::Json& jsonDefinition,
-                      std::shared_ptr<const store::IStoreRead> storeRead,
+                      std::shared_ptr<store::IStoreRead> storeRead,
```
Comment (juliancnn) [inline src/engine/source/builder/policy.hpp:272]: Why did the IStoreRead stop being constant if it is a reader interface?

File: src/engine/source/store/interface/store/utils.hpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+#define _STORE_UTILS_H
+
+#include <memory>
+#include <optional>
+#include <variant>
+
+#include "store/istore.hpp"
+#include <error.hpp>
+#include <json/json.hpp>
+#include <name.hpp>
```
Comment (juliancnn) [inline src/engine/source/store/interface/store/utils.hpp:11]:
```suggestion
#include <error.hpp>
#include <json/json.hpp>
#include <name.hpp>

#include <store/istore.hpp>
```

File: src/engine/source/store/interface/store/utils.hpp [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    auto wrappedContent = json::Json {};
+    wrappedContent.setObject();
+
+    // Save the original json and yml
+    wrappedContent.set("/json", contentJson);
+    wrappedContent.setString(original, "/original");
+    wrappedContent.setString(format, "/format");
+
+    return wrappedContent;
+}
```
Comment (juliancnn) [inline src/engine/source/store/interface/store/utils.hpp:43]: The store must not know the Asset, nor its terminology,

File: src/engine/source/store/interface/store/utils.hpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+ * This function retrieves data from the provided store using the specified name.
+ * Optionally, if 'original' is set to true, it returns the original data.
+ *
+ * @param storeRead The store to retrieve data from.
+ * @param name The name of the data to retrieve.
+ * @param original Flag to indicate whether to retrieve the original data.
+ * @return A variant containing the retrieved JSON data or an error.
+ */
+static std::variant<json::Json, base::Error>
+get(std::shared_ptr<store::IStoreRead> storeRead, const base::Name& name, bool original = false)
```
Comment (juliancnn) [inline src/engine/source/store/interface/store/utils.hpp:57]:
```suggestion
get(const std::shared_ptr<store::IStoreRead>& storeRead, const base::Name& name, bool original = false)
```

File: src/engine/source/store/interface/store/utils.hpp [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+#include "store/istore.hpp"
+#include <error.hpp>
+#include <json/json.hpp>
+#include <name.hpp>
+
+/**
+ * @brief Store functionallity.
+ *
+ */
+namespace store
```
Comment (juliancnn) [inline src/engine/source/store/interface/store/utils.hpp:17]: Is a different approach possible for these functions? I find it confusing to have:
`store->get(xx)` and `store::get(store, xx)`.

File: src/engine/source/store/interface/store/utils.hpp [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        return json;
+    }
+
+    auto jsonValue = json.getJson("/json");
+    if (!jsonValue.has_value())
+    {
+        return base::Error {"/json path not found in JSON."};
+    }
+
+    return jsonValue.value();
```
Comment (juliancnn) [inline src/engine/source/store/interface/store/utils.hpp:82]: move

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 14]
```diff
@@ -323,33 +329,64 @@ std::optional<base::Error> Catalog::putResource(const Resource& item, const std:
...
-    const auto formatResult = formatterIt->second(contentJson);
-    if (std::holds_alternative<base::Error>(formatResult))
+
+    if (Resource::formatToStr(Format::json) == content.getString("/format").value())
     {
-        return base::Error {fmt::format("JSON object could not be created from '{} {}': {}",
-                                        Resource::formatToStr(resource.m_format),
-                                        resource.m_name.fullName(),
-                                        std::get<base::Error>(formatResult).message)};
+        if (Resource::formatToStr(Format::yaml) == Resource::formatToStr(resource.m_format))
```
Comment (juliancnn) [inline src/engine/source/api/src/catalog/catalog.cpp:383]: unify the `if` confitions:
```suggestion
        if (Resource::formatToStr(Format::json) == Resource::formatToStr(content.getString("/format").value())
        && Resource::formatToStr(Format::yaml) == Resource::formatToStr(resource.m_format))
```

**Review by juliancnn — CHANGES_REQUESTED** [review]

**Review by juliancnn — APPROVED** [review]
LGTM!

File: src/engine/source/api/src/catalog/catalog.cpp [hunk 15]
```diff
@@ -246,6 +250,8 @@ std::optional<base::Error> Catalog::postResource(const Resource& collection, con
 
 std::optional<base::Error> Catalog::putResource(const Resource& item, const std::string& content)
 {
+    using Format = ::com::wazuh::api::engine::catalog::ResourceFormat;
```
Comment (soynof) [inline src/engine/source/api/src/catalog/catalog.cpp:253]: Could it be that this resource is not being used?

File: src/engine/tools/engine-suite/src/engine_clear/__init__.py
[the remark this answers is not in the document]
Comment (NahuFigueroa97) [reply src/engine/tools/engine-suite/src/engine_clear/__init__.py:1]: Uploading the __init__.py file to the repository is a good practice. This ensures that anyone cloning or downloading the repository gets all the files needed for the code to work smoothly in their local environment.

File: src/engine/source/api/src/catalog/catalog.cpp (hunk 15 above)
Comment (NahuFigueroa97) [reply src/engine/source/api/src/catalog/catalog.cpp]: good catch!

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Pull request merged**